crates/fuel-streams-core/src/stream.rs

```diff
@@ -106,27 +104,36 @@ impl<S: Streamable> Stream<S> {
         &self,
         subject: &impl IntoSubject,
         payload: &S,
-    ) -> Result<u64, StreamError> {
+    ) -> Result<i64, StreamError> {
         let subject_name = &subject.parse();
-        self.store
-            .put(subject_name, payload.encode(subject_name).await.into())
-            .await
-            .map_err(|s| StreamError::PublishFailed {
+        let result = self
+            .store
+            .create(subject_name, payload.encode(subject_name).await.into())
+            .await;
+
+        match result {
+            Ok(sequence) => Ok(sequence as i64),
+            Err(e) if e.kind() == CreateErrorKind::AlreadyExists => {
+                // This is a workaround to avoid publish two times the same message
+                Ok(-1)
+            }
+            Err(e) => Err(StreamError::PublishFailed {
                 subject_name: subject_name.to_string(),
-                source: s,
-            })
+                source: e,
+            }),
+        }
     }
 
     pub async fn subscribe(
         &self,
         // TODO: Allow encapsulating Subject to return wildcard token type
         wildcard: &str,
-    ) -> Result<impl futures::Stream<Item = Vec<u8>>, StreamError> {
-        Ok(self
-            .store
-            .watch(&wildcard)
-            .await
-            .map(|stream| stream.map(|entry| entry.unwrap().value.to_vec()))?)
+    ) -> Result<impl futures::Stream<Item = Option<Vec<u8>>>, StreamError> {
```

Comment:
Why would we want this?

```suggestion
    ) -> Result<impl futures::Stream<Item = Vec<u8>>, StreamError> {
```

A stream should always yield the published bytes and there should never be a situation where we publish nothing (`None`). What do you think?

Reply:
I you check what the entry value is in the closure you will see its a Result with a WatcherErrorKind error. There is two ways we can think here: 1. if error we do an Option and return that not revealing the details to the user. The other option would be to actually have return `Result<impl futures::Stream<Item =Result< Vec<u8>>, WatcherErrorKind>, StreamError>` to the user and let him deal with the error if it occurs. Since we mentioned tha twe dont want to expose the internal details of our impl (here a kv store), we better conceal that error and just map it to an Option. Many other systems do that. The user normally has to do: while let Some(x) = stream.await { ... }. This was my reasoning here. Also, when testing, killing the nats server instantaneously results in WatcherKindError. Previously we were unwrapping it which caused a bad panic. Here is another reason why we want to handle this either 1 or 2 above.

Reply:
Thanks for the detailed explanation 👍🏾